File: backend/djadmin/monitor/log_management.py
"""日志存储（OpenSearch）侧的管理面构建器。

对应《日志采集架构文档》第 4/5 节：索引命名、index template、ISM policy、
ingest pipeline 的默认体。全部为纯函数，不触碰网络，便于单测；
bootstrap_log_storage 是唯一执行写入的编排入口。
"""
import re

from django.db import transaction

from .log_schema import STANDARD_LOG_FIELDS
from .opensearch_client import OpenSearchClient, OpenSearchError

_INDEX_SEGMENT_RE = re.compile(r'[^a-z0-9_-]+')


def _safe_segment(value):
    """索引名段只允许小写字母/数字/下划线/连字符，避免非法索引名。"""
    segment = _INDEX_SEGMENT_RE.sub('-', str(value or '').strip().lower()).strip('-')
    return segment or 'unknown'


def build_index_name(index_prefix, project_code, environment_code, business_system_code, tier_code):
    """logs-<project.code>-<environment.code>-<business_system.code>-<tier.code>（§4.1）。

    服务/实例/主机/日志类型不进索引名，作为字段存储。
    project 和 tier 都直接决定索引归属，缺失时必须报错，不能静默写成 unknown 而造出难以清理的 data stream。
    """
    if not str(project_code or '').strip():
        raise ValueError('缺少所属项目')
    if not str(tier_code or '').strip():
        raise ValueError('缺少保留档位')
    return '-'.join([
        _safe_segment(index_prefix or 'logs'),
        _safe_segment(project_code),
        _safe_segment(environment_code),
        _safe_segment(business_system_code),
        _safe_segment(tier_code),
    ])


def build_index_template_name(index_prefix):
    return f'{_safe_segment(index_prefix or "logs")}-template'


def build_index_template_body(index_prefix):
    """data stream 模板，单分片、限制字段总数（§4.4 / §12 mapping 膨胀）。"""
    prefix = _safe_segment(index_prefix or 'logs')
    return {
        'index_patterns': [f'{prefix}-*'],
        'data_stream': {},
        'template': {
            'settings': {
                'number_of_shards': 1,
                'number_of_replicas': 0,
                'index.refresh_interval': '10s',
                'index.mapping.total_fields.limit': 2000,
            },
            'mappings': {
                # 标准字段之外一律不再自动建 mapping，应用差异字段必须写进 app_fields。
                'dynamic': False,
                'properties': STANDARD_LOG_FIELDS,
            },
        },
    }


def build_ism_policy_name(index_prefix, tier_code):
    return f'{_safe_segment(index_prefix or "logs")}-{_safe_segment(tier_code)}-retention'


def build_ism_policy_body(index_prefix, tier):
    """按档位记录生成 ISM policy，经 ism_template 按索引名后缀自动挂载（§4.5）。

    tier 是 monitor.LogRetentionTier 实例：档位已改为用户可维护的数据。
    """
    prefix = _safe_segment(index_prefix or 'logs')
    code = _safe_segment(tier.code)
    return {
        'policy': {
            'description': (
                f'{prefix} *-{code} 索引保留 {tier.retention_days} 天'
                f'（预估 {tier.daily_size_gb}GB/天，合计 {tier.estimated_total_gb}GB）'
            ),
            'default_state': 'hot',
            'states': [
                {
                    'name': 'hot',
                    'actions': [
                        {
                            'rollover': {
                                'min_primary_shard_size': tier.rollover_min_primary_shard_size,
                                'min_index_age': tier.rollover_min_index_age,
                            },
                        },
                    ],
                    'transitions': [
                        {
                            'state_name': 'delete',
                            'conditions': {'min_index_age': f'{tier.retention_days}d'},
                        },
                    ],
                },
                {'name': 'delete', 'actions': [{'delete': {}}]},
            ],
            'ism_template': [{'index_patterns': [f'{prefix}-*-{code}'], 'priority': 100}],
        },
    }


def _upsert_ism_policy(client, name, body):
    """覆盖已存在的 ISM policy 必须带乐观锁参数，否则 OpenSearch 直接返回 409。"""
    try:
        existing = client.get_ism_policy(name)
    except OpenSearchError:
        existing = None
    params = None
    if existing and existing.get('_seq_no') is not None:
        params = {'if_seq_no': existing['_seq_no'], 'if_primary_term': existing['_primary_term']}
    client.put_ism_policy(name, body, params=params)


def build_pipeline_name(application_code, log_name):
    """app-<application.code>-<log_name>：解析规则跟随应用类型，不随实例增长（§5.2）。"""
    return f'app-{_safe_segment(application_code)}-{_safe_segment(log_name)}'


def bootstrap_log_storage(cluster):
    """确保 index template 与各保留档位的 ISM policy 存在（§11 阶段 1 的自动化部分）。

    幂等：重复执行只会覆盖同名对象，不重复创建。返回写入的对象名清单。
    档位新增/修改后需重新执行本方法，策略才会下发到集群。
    """
    from .models import LogRetentionTier

    client = OpenSearchClient(cluster)
    prefix = cluster.index_prefix or 'logs'

    template_name = build_index_template_name(prefix)
    client.put_index_template(template_name, build_index_template_body(prefix))

    policies = []
    for tier in LogRetentionTier.objects.filter(enabled=True):
        policy_name = build_ism_policy_name(prefix, tier.code)
        _upsert_ism_policy(client, policy_name, build_ism_policy_body(prefix, tier))
        policies.append(policy_name)

    return {'index_template': template_name, 'ism_policies': policies}


def sync_log_storage_quietly(cluster):
    """下发模板与保留策略，失败只返回原因、不抛异常。

    索引模板是代码派生物而非用户配置，必须自动跟随代码版本；但集群暂时不可达
    不应阻塞集群保存或数据库迁移，因此这里吞掉异常，由调用方决定如何提示。
    """
    if not cluster or not cluster.enabled:
        return ''
    try:
        bootstrap_log_storage(cluster)
    except OpenSearchError as exc:
        return str(exc)[:500]
    return ''


def enqueue_log_storage_sync(cluster_id):
    """在事务提交后入队，避免 worker 读取到尚未提交的集群配置。"""
    from .models import OpenSearchCluster
    from .tasks import sync_log_storage

    cluster = OpenSearchCluster.objects.filter(pk=cluster_id, enabled=True).first()
    if cluster is None:
        return
    cluster.storage_sync_status = OpenSearchCluster.StorageSyncStatus.PENDING
    cluster.storage_sync_error = ''
    cluster.storage_sync_time = None
    cluster.save(update_fields=['storage_sync_status', 'storage_sync_error', 'storage_sync_time', 'update_time'])

    # Celery Task.delay 的运行时 API 正确，Pylance 对 shared_task 返回对象的存根误判为列表。
    transaction.on_commit(lambda: sync_log_storage.delay(cluster_id))  # type: ignore[operator]

File: autoadmin/internal/monitor/log_management.go
package monitor

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
	"time"

	db "autoadmin/internal/platform/database/generated"

	"github.com/gin-gonic/gin"
)

// Elasticsearch 日志存储管理面：索引命名、index template、ILM policy、ingest pipeline
// 的期望态构建与下发编排，对应 Django monitor/log_management.py。
// 构建器全部为纯函数便于单测；bootstrapElasticsearchStorage 是唯一执行写入的编排入口。

var indexSegmentPattern = regexp.MustCompile(`[^a-z0-9_-]+`)

// safeIndexSegment 索引名段只允许小写字母/数字/下划线/连字符，避免非法索引名。
func safeIndexSegment(value string) string {
	segment := strings.Trim(indexSegmentPattern.ReplaceAllString(strings.ToLower(strings.TrimSpace(value)), "-"), "-")
	if segment == "" {
		return "unknown"
	}
	return segment
}

func buildIndexTemplateName(indexPrefix string) string {
	return safeIndexSegment(indexPrefix) + "-template"
}

// safeOptionalSegment 与 safeIndexSegment 类似，但空值返回空串（不兜底 unknown），
// 便于调用方自行决定默认段。
func safeOptionalSegment(value string) string {
	return strings.Trim(indexSegmentPattern.ReplaceAllString(strings.ToLower(strings.TrimSpace(value)), "-"), "-")
}

// processingPipelineName ES ingest pipeline id：<索引前缀(默认 autoadmin)>-<应用 code 或 general>-<规则名>。
// 采集端 Filebeat 的 pipeline 字段、后端发布/删除、链路体检都必须用同一个名字，禁止各自拼接。
func processingPipelineName(prefix, application, rule string) string {
	prefixSegment := safeOptionalSegment(prefix)
	if prefixSegment == "" {
		prefixSegment = "autoadmin"
	}
	applicationSegment := safeOptionalSegment(application)
	if applicationSegment == "" {
		applicationSegment = "general"
	}
	ruleSegment := safeOptionalSegment(rule)
	if ruleSegment == "" {
		ruleSegment = "rule"
	}
	return strings.Join([]string{prefixSegment, applicationSegment, ruleSegment}, "-")
}

// buildIndexTemplateBody data stream 模板：单分片、限制字段总数、标准字段之外不再自动建 mapping。
// 字段契约与 Django log_schema.STANDARD_LOG_FIELDS 一致。
func buildIndexTemplateBody(indexPrefix string) gin.H {
	prefix := safeIndexSegment(indexPrefix)
	return gin.H{
		"index_patterns": []string{prefix + "-*"},
		"data_stream":    gin.H{},
		"template": gin.H{
			"settings": gin.H{
				"number_of_shards":                 1,
				"number_of_replicas":               0,
				"index.refresh_interval":           "10s",
				"index.mapping.total_fields.limit": 2000,
			},
			"mappings": gin.H{
				"dynamic":    false,
				"properties": standardLogFields,
			},
		},
	}
}

var standardLogFields = gin.H{
	"@timestamp":        gin.H{"type": "date"},
	"message":           gin.H{"type": "text"},
	"project":           gin.H{"type": "keyword"},
	"business_system":   gin.H{"type": "keyword"},
	"environment":       gin.H{"type": "keyword"},
	"service":           gin.H{"type": "keyword"},
	"application":       gin.H{"type": "keyword"},
	"instance":          gin.H{"type": "keyword"},
	"host_ip":           gin.H{"type": "keyword"},
	"log_name":          gin.H{"type": "keyword"},
	"log_path":          gin.H{"type": "keyword"},
	"log_level":         gin.H{"type": "keyword"},
	"log_time":          gin.H{"type": "keyword"},
	"log_message":       gin.H{"type": "text"},
	"error_fingerprint": gin.H{"type": "keyword"},
	// flat_object 是 Elasticsearch 专有类型；Elasticsearch 用 flattened（8.x 自带，无需额外插件）。
	"app_fields": gin.H{"type": "flattened"},
}

func buildILMPolicyName(indexPrefix, tierCode string) string {
	return safeIndexSegment(indexPrefix) + "-" + safeIndexSegment(tierCode) + "-retention"
}

// buildTierIndexTemplateName 每档位一个索引模板，负责把该档位的 ILM 策略绑到匹配的 data stream。
// ELM/ILM 没有 ILM 的 ism_template 自动挂载，只能在索引模板里写 index.lifecycle.name。
func buildTierIndexTemplateName(indexPrefix, tierCode string) string {
	return safeIndexSegment(indexPrefix) + "-" + safeIndexSegment(tierCode) + "-template"
}

type retentionTierRow struct {
	Code                string
	RetentionDays       int64
	DailySizeGB         float64
	RolloverMinIndexAge string
}

// buildILMPolicyBody 按档位生成 Elasticsearch ILM policy。
// 滚动阈值不小于 1gb（避免小档位产生大量碎索引）；删除阶段 min_age = 保留天数。
func buildILMPolicyBody(indexPrefix string, tier retentionTierRow) gin.H {
	rolloverSize := fmt.Sprintf("%dgb", int(math.Max(1, math.Round(tier.DailySizeGB))))
	rollover := gin.H{"max_primary_shard_size": rolloverSize}
	if age := strings.TrimSpace(tier.RolloverMinIndexAge); age != "" {
		rollover["max_age"] = age
	}
	return gin.H{"policy": gin.H{
		"phases": gin.H{
			"hot": gin.H{
				"actions": gin.H{"rollover": rollover},
			},
			"delete": gin.H{
				"min_age": fmt.Sprintf("%dd", tier.RetentionDays),
				"actions": gin.H{"delete": gin.H{}},
			},
		},
	}}
}

// buildTierIndexTemplateBody 在基础模板之上叠加「按档位后缀匹配 + 绑定 ILM 策略」。
// 必须是自包含模板（同时带 data_stream/settings/mappings）：实测 ES 8.13 下高优先级模板
// 与基础模板重叠时不会把基础模板的 mappings/settings 合并进 data stream，缺了会丢字段映射。
func buildTierIndexTemplateBody(indexPrefix string, tier retentionTierRow) gin.H {
	prefix := safeIndexSegment(indexPrefix)
	code := safeIndexSegment(tier.Code)
	body := buildIndexTemplateBody(indexPrefix)
	body["index_patterns"] = []string{fmt.Sprintf("%s-*-%s", prefix, code)}
	body["priority"] = 200
	settings := body["template"].(gin.H)["settings"].(gin.H)
	settings["index.lifecycle.name"] = buildILMPolicyName(indexPrefix, tier.Code)
	return body
}

// pipelineSignature 只取行为字段：processors 有序直接序列化，描述等非行为字段不参与比对。
// encoding/json 对 map 按 key 排序输出，与 Django json.dumps(sort_keys=True) 一致。
func pipelineSignature(body any) string {
	var decoded map[string]any
	switch typed := body.(type) {
	case map[string]any:
		decoded = typed
	case gin.H:
		decoded = typed
	case json.RawMessage:
		_ = json.Unmarshal(typed, &decoded)
	case []byte:
		_ = json.Unmarshal(typed, &decoded)
	}
	processors, onFailure := []any{}, []any{}
	if decoded != nil {
		if items, ok := decoded["processors"].([]any); ok {
			processors = items
		}
		if items, ok := decoded["on_failure"].([]any); ok {
			onFailure = items
		}
	}
	encoded, _ := json.Marshal(gin.H{"processors": processors, "on_failure": onFailure})
	return string(encoded)
}

// indexPatternsOverlap 判断两个索引 pattern 是否可能匹配同一索引：去掉通配符后互为前缀即视为重叠。
func indexPatternsOverlap(left, right string) bool {
	a := strings.ReplaceAll(strings.TrimSpace(left), "*", "")
	b := strings.ReplaceAll(strings.TrimSpace(right), "*", "")
	if a == "" || b == "" {
		return true
	}
	return strings.HasPrefix(a, b) || strings.HasPrefix(b, a)
}

// splitCatPatterns 解析 _cat/templates 的 index_patterns（形如 "[logs*, logs-*]"）为列表。
func splitCatPatterns(raw string) []string {
	raw = strings.TrimSpace(raw)
	raw = strings.TrimPrefix(raw, "[")
	raw = strings.TrimSuffix(raw, "]")
	result := []string{}
	for _, pattern := range strings.Split(raw, ",") {
		if pattern = strings.TrimSpace(pattern); pattern != "" {
			result = append(result, pattern)
		}
	}
	return result
}

// conflictingIndexTemplates 查同 priority 且 index_patterns 与本模板重叠的已有模板（排除自身）。
// ES 本身也会 400，但错误体很长；提前给出简明报错，让用户直接删掉冲突模板或调整其 priority。
func (handler *Handler) conflictingIndexTemplates(context *gin.Context, cluster elasticsearchCluster, selfName string, patterns []string, priority int) []string {
	rows, err := handler.elasticsearchRequestArray(context, cluster, "GET", "/_cat/templates?format=json&h=name,index_patterns,priority")
	if err != nil {
		return nil
	}
	conflicts := []string{}
	for _, row := range rows {
		name := catString(row, "name")
		if name == "" || name == selfName {
			continue
		}
		candidatePriority := 0
		if raw, ok := row["priority"]; ok {
			candidatePriority, _ = strconv.Atoi(strings.TrimSpace(fmt.Sprint(raw)))
		}
		if candidatePriority != priority {
			continue
		}
		for _, candidate := range splitCatPatterns(catString(row, "index_patterns")) {
			for _, want := range patterns {
				if indexPatternsOverlap(candidate, want) {
					conflicts = append(conflicts, fmt.Sprintf("%s (index_patterns=%s, priority=%d)", name, candidate, priority))
				}
			}
		}
	}
	return conflicts
}

// bootstrapElasticsearchStorage 确保基础索引模板、各档位的 ILM policy 与档位索引模板存在（幂等）。
func (handler *Handler) bootstrapElasticsearchStorage(context *gin.Context, cluster elasticsearchCluster) error {
	prefix := cluster.IndexPrefix
	if strings.TrimSpace(prefix) == "" {
		prefix = "autoadmin"
	}
	// 模板名必须是 `<prefix>-template`（与 Django / LOG_COLLECTION_ARCHITECTURE §4.4 一致）。
	// 早期版本误用了不带后缀的 `<prefix>`，导致健康检查查 `logs-template` 却报了另一个名字；
	// 这里同时清理那个错名模板，避免同一 index_patterns 挂着两份模板。
	baseName, basePatterns := buildIndexTemplateName(prefix), []string{prefix + "-*"}
	if conflicts := handler.conflictingIndexTemplates(context, cluster, baseName, basePatterns, 0); len(conflicts) > 0 {
		return fmt.Errorf("存在与索引模板 %s（priority=0）冲突的模板：%s；请删除该模板或调整其 priority 后重试",
			baseName, strings.Join(conflicts, "; "))
	}
	if _, err := handler.elasticsearchRequest(context, cluster, "PUT", "/_index_template/"+baseName, buildIndexTemplateBody(prefix)); err != nil {
		return err
	}
	handler.cleanupLegacyIndexTemplate(context, cluster, prefix)
	tiers, err := handler.loadEnabledRetentionTiers(context)
	if err != nil {
		return err
	}
	for _, tier := range tiers {
		// PUT _ilm/policy 是幂等覆盖（不像 ISM 需要 if_seq_no 乐观锁）。
		if _, err := handler.elasticsearchRequest(context, cluster, "PUT", "/_ilm/policy/"+buildILMPolicyName(prefix, tier.Code), buildILMPolicyBody(prefix, tier)); err != nil {
			return err
		}
		// 档位模板负责把 ILM 策略绑到 `logs-*-<tier>` 的 data stream。
		tierName, tierPatterns := buildTierIndexTemplateName(prefix, tier.Code), []string{fmt.Sprintf("%s-*-%s", safeIndexSegment(prefix), safeIndexSegment(tier.Code))}
		if conflicts := handler.conflictingIndexTemplates(context, cluster, tierName, tierPatterns, 200); len(conflicts) > 0 {
			return fmt.Errorf("存在与索引模板 %s（priority=200）冲突的模板：%s；请删除该模板或调整其 priority 后重试",
				tierName, strings.Join(conflicts, "; "))
		}
		if _, err := handler.elasticsearchRequest(context, cluster, "PUT", "/_index_template/"+tierName, buildTierIndexTemplateBody(prefix, tier)); err != nil {
			return err
		}
	}
	return nil
}

// cleanupLegacyIndexTemplate 删除历史版本用错名字（`<prefix>` 而非 `<prefix>-template`）建的模板。
// 只在该名字确实存在、且 index_patterns 命中 `<prefix>-*` 时才删，避免误删恰好同名的其它模板。
// 失败不影响 bootstrap 结果（`logs-template` 已写入，且同优先级下后建者生效）。
func (handler *Handler) cleanupLegacyIndexTemplate(context *gin.Context, cluster elasticsearchCluster, prefix string) {
	legacy := safeIndexSegment(prefix)
	if legacy == buildIndexTemplateName(prefix) {
		return
	}
	existing, err := handler.elasticsearchRequest(context, cluster, "GET", "/_index_template/"+legacy, nil)
	if err != nil || !indexTemplateMatchesPrefix(existing, legacy) {
		return
	}
	_, _ = handler.elasticsearchRequest(context, cluster, "DELETE", "/_index_template/"+legacy, nil)
}

// indexTemplateMatchesPrefix 判断 GET /_index_template/<name> 的响应里 pattern 是否正好是 `<prefix>-*`。
func indexTemplateMatchesPrefix(payload map[string]any, prefix string) bool {
	templates, _ := payload["index_templates"].([]any)
	if len(templates) == 0 {
		return false
	}
	first, _ := templates[0].(map[string]any)
	definition, _ := first["index_template"].(map[string]any)
	patterns, _ := definition["index_patterns"].([]any)
	want := prefix + "-*"
	for _, pattern := range patterns {
		if fmt.Sprint(pattern) == want {
			return true
		}
	}
	return false
}

func (handler *Handler) loadEnabledRetentionTiers(context *gin.Context) ([]retentionTierRow, error) {
	rows, err := db.New(handler.db).ListEnabledRetentionTiers(context)
	if err != nil {
		return nil, err
	}
	tiers := make([]retentionTierRow, 0, len(rows))
	for _, row := range rows {
		tiers = append(tiers, retentionTierRow{
			Code: row.Code, RetentionDays: int64(row.RetentionDays), DailySizeGB: row.DailySizeGb,
			RolloverMinIndexAge: row.RolloverMinIndexAge,
		})
	}
	return tiers, nil
}

// syncClusterLogStorage 对单个启用集群下发模板与保留策略，并回写 storage_sync_* 状态，
// 对应 Django 的 celery 任务 sync_log_storage（失败不抛出，只落状态供前端展示）。
func (handler *Handler) syncClusterLogStorage(clusterID int64) {
	ginContext, _ := gin.CreateTestContext(nil)
	queries := db.New(handler.db)
	err := queries.MarkClusterStorageSyncPending(ginContext, db.MarkClusterStorageSyncPendingParams{
		UpdateTime: time.Now().UTC(), ID: clusterID,
	})
	if err != nil {
		return
	}
	cluster, err := handler.loadElasticsearchClusterByID(ginContext, clusterID)
	fail := func(message string) {
		_ = queries.MarkClusterStorageSyncFailed(ginContext, db.MarkClusterStorageSyncFailedParams{
			StorageSyncError: message, StorageSyncTime: sql.NullTime{Time: time.Now().UTC(), Valid: true},
			UpdateTime: time.Now().UTC(), ID: clusterID,
		})
	}
	switch {
	case err != nil:
		fail(err.Error())
	case !cluster.Enabled:
		fail("集群未启用，跳过同步")
	default:
		if bootstrapErr := handler.bootstrapElasticsearchStorage(ginContext, cluster); bootstrapErr != nil {
			fail(truncateElasticsearchError(bootstrapErr))
		} else {
			_ = queries.MarkClusterStorageSyncSuccess(ginContext, db.MarkClusterStorageSyncSuccessParams{
				StorageSyncTime: sql.NullTime{Time: time.Now().UTC(), Valid: true},
				UpdateTime:      time.Now().UTC(), ID: clusterID,
			})
		}
	}
}

// syncAllClusterLogStorage 对所有启用集群异步下发，对应 Django 档位改动后的 _apply_policies。
func (handler *Handler) syncAllClusterLogStorage() {
	ginContext, _ := gin.CreateTestContext(nil)
	ids, err := db.New(handler.db).ListEnabledElasticsearchClusterIDs(ginContext)
	if err != nil {
		return
	}
	for _, id := range ids {
		handler.syncClusterLogStorage(id)
	}
}

func truncateElasticsearchError(err error) string {
	message := err.Error()
	if len(message) > 500 {
		return message[:500]
	}
	return message
}

// publishProcessingPipeline 把解析规则的 pipeline 发布到集群（Django 规则增改的同步行为）。
func (handler *Handler) publishProcessingPipeline(context *gin.Context, clusterID int64, name string, pipelineBody any) error {
	cluster, err := handler.loadElasticsearchClusterByID(context, clusterID)
	if err != nil {
		return err
	}
	if !cluster.Enabled {
		return fmt.Errorf("Elasticsearch 集群未启用")
	}
	if _, err := handler.elasticsearchRequest(context, cluster, "PUT", "/_ingest/pipeline/"+name, pipelineBody); err != nil {
		return err
	}
	return nil
}

func (handler *Handler) deleteProcessingPipeline(context *gin.Context, clusterID int64, name string) error {
	cluster, err := handler.loadElasticsearchClusterByID(context, clusterID)
	if err != nil {
		return err
	}
	if !cluster.Enabled {
		return nil
	}
	if _, err := handler.elasticsearchRequest(context, cluster, "DELETE", "/_ingest/pipeline/"+name, nil); err != nil {
		return err
	}
	return nil
}
